### ISSUE TITLE
PayPal Payflow Pro performs delayed capture with incorrect parent transaction ID

### ISSUE THREAD

**Issue by aaronsilber** [magento/magento2#30870]
Shortly after updating to Magento 2.3.6, our client had a large volume of customers report they had been "charged twice." After investigating this, we found these customers had a lingering authorization (pending transaction) on their cards, in addition to a charge for the full amount of the order. The authorization remains until it expires, typically 30 days.

We found that [changes to the Magento_Paypal module introduced in Magento 2.3.6](https://github.com/magento/magento2/commit/24481e96c171a51ca326f107ca76662bdec8732d#diff-7198862901baba9a4936f420cf43203f92e1e221c341489f966eb69467f163ebR206) caused the incorrect parent transaction ID to be used when performing a delayed capture. Instead of using the order's authorization as the parent transaction ID (behavior prior to Magento 2.3.6), the $0.00 vault token authorization for the payment is used as the parent transaction ID.

This causes the authorization to remain open until it expires, and the transaction sent to Payflow is not associated to the original order authorization. Additionally, the transaction displays in Payflow as "Capture" instead of "Delayed Capture". 

Merchants with "Payment Action" set to "Sale" are likely not impacted, but we have not tested this.

### Preconditions (*)

1. Magento Open Source >= 2.3.6
3. magento/module-paypal >= 100.3.6

### Steps to reproduce (*)

1. Set up PayPal Payflow Pro gateway with "Payment Action" set to "Authorization"
2. Place an order with a credit card
3. Capture the order's payment by creating an invoice and choosing to Capture Online

### Expected result (*)

![payflow-transaction-correct-delayed-capture](https://user-images.githubusercontent.com/233446/98734014-65ba9200-236f-11eb-8c75-2d41dc087cd5.png)
1. Magento Transaction is created with parent transaction ID of the authorization created at checkout
2. A "Delayed Capture" transaction is sent to Payflow with a parent transaction ID of the authorization created at checkout
3. Customer observes a new charge on their credit card for the amount of the invoice
4. Customer observes pending transaction disappear from their card immediately, or within several days

### Actual result (*)

![payflow-transaction-incorrect-capture](https://user-images.githubusercontent.com/233446/98734065-71a65400-236f-11eb-9c7b-84dfea977287.png)
1. Magento Transaction is created with parent transaction ID of the authorization created at checkout
2. A **"Sale"** transaction is sent to Payflow with a parent transaction ID of the **vault token created at checkout**
3. Customer observes a new charge on their credit card for the amount of the invoice
4. **Customer observes pending transaction remain on their card for over a week, and until it expires, typically in 30 days**
5. Customer contacts merchant and reports they have been "charged twice"

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [x] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by aaronsilber** [comment]
To fix this for our client, we created a Composer patch that partially reverts the [modifications to the parent transaction IDs used in authorize and capture](https://github.com/magento/magento2/commit/24481e96c171a51ca326f107ca76662bdec8732d). This patch can be applied by creating a directory "patches", and extracting the attached file to patches/MAGENTO-TWO-30870.patch, and adding the following to composer.json:

```
        "patches": {
            "magento/module-paypal": {
                "Prevent duplicate authorizations when using Payflow with Delayed Capture": "patches/MAGENTO-TWO-30870.patch"
            }
        }
```

I'm working on reformatting this as a PR and will attach it once it's ready.

[MAGENTO-TWO-30870.zip](https://github.com/magento/magento2/files/5520288/MAGENTO-TWO-30870.zip)

**Comment by aaronsilber** [comment]
@magento give me 2.4-develop instance

**Comment by aaronsilber** [comment]
Reproduced this issue on fresh 2.4-develop instance in Payflow Test Mode.

Note how the Magento Transaction displays the correct parent transaction ID for the invoice capture, but reviewing the actual transaction data in Payflow shows the incorrect parent transaction ID was used when submitting the transaction data to Payflow.

Magento transaction data after placing an order:
![m2_30870_1_order_auth_tx](https://user-images.githubusercontent.com/233446/98858985-7b8f8c00-242f-11eb-8e67-f3a28c101f8a.png)

Payflow transaction data after placing an order:
![m2_30870_2_order_auth_tx_payflow](https://user-images.githubusercontent.com/233446/98859059-97932d80-242f-11eb-9ef4-ff85ebe3ad75.png)

Magento transaction list after capturing with an invoice:
![m2_30870_3_order_capture_tx_list](https://user-images.githubusercontent.com/233446/98859128-aed21b00-242f-11eb-94e5-0f839c94ac7b.png)

Magento transaction data for the capture (shows expected parent transaction ID):
![m2_30870_4_order_capture_tx](https://user-images.githubusercontent.com/233446/98859170-c01b2780-242f-11eb-828f-c1576fe30832.png)

Payflow transaction data for the capture; shows incorrect parent transaction ID as well as transaction type "Sale" instead of expected "Delayed Capture":
![m2_30870_5_order_capture_tx_payflow](https://user-images.githubusercontent.com/233446/98859244-de812300-242f-11eb-918d-1a2886ee5890.png)

**Comment by aaronsilber** [comment]
@magento I am working on this

**Comment by aaronsilber** [comment]
@magento add to contributors team

**Comment by aaronsilber** [comment]
@magento I am working on this

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @sdzhepa 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-42279` were created

**Issue Available**: @sdzhepa, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by gabrieldagama** [comment]
Hi @aaronsilber. Thank you for your report. 
The issue has been fixed in magento/magento2#30885 by @aaronsilber in 2.4-develop branch
Related commit(s):
 - [fde3fd7bc0964b60f841c9f6f484ba70f836f087](https://github.com/magento/magento2/commit/fde3fd7bc0964b60f841c9f6f484ba70f836f087)
 - [0d79d7d89730cc6ee557fb9893d90f822a3ec6e7](https://github.com/magento/magento2/commit/0d79d7d89730cc6ee557fb9893d90f822a3ec6e7)
 - [7c43141f7438551fc4b07d0154109daba8b6308a](https://github.com/magento/magento2/commit/7c43141f7438551fc4b07d0154109daba8b6308a)
 - [89c2e12703b595a06414955dc1df6b1b31d4c901](https://github.com/magento/magento2/commit/89c2e12703b595a06414955dc1df6b1b31d4c901)
 - [c00fad647a87f173b8729dcefd89c6dd4bed15d3](https://github.com/magento/magento2/commit/c00fad647a87f173b8729dcefd89c6dd4bed15d3)
 - [3a89dfa6532180eb80ae58184d6f1fb57e4f4fc7](https://github.com/magento/magento2/commit/3a89dfa6532180eb80ae58184d6f1fb57e4f4fc7)
 - [dbbcfc470277afb02b9558d93aa98cb8b7f67fcc](https://github.com/magento/magento2/commit/dbbcfc470277afb02b9558d93aa98cb8b7f67fcc)

The fix will be available with the upcoming 2.4.3 release.

**Comment by 0xMatt** [comment]
Not fixed in 2.4.3-p1. 

When I capture a payment, a sales transaction is made. This doesn't close out the authorization and results in a double charge.

**Comment by 0xMatt** [comment]
> To fix this for our client, we created a Composer patch that partially reverts the [modifications to the parent transaction IDs used in authorize and capture](https://github.com/magento/magento2/commit/24481e96c171a51ca326f107ca76662bdec8732d). This patch can be applied by creating a directory "patches", and extracting the attached file to patches/MAGENTO-TWO-30870.patch, and adding the following to composer.json:
> 
> ```
>         "patches": {
>             "magento/module-paypal": {
>                 "Prevent duplicate authorizations when using Payflow with Delayed Capture": "patches/MAGENTO-TWO-30870.patch"
>             }
>         }
> ```
> 
> I'm working on reformatting this as a PR and will attach it once it's ready.
> 
> [MAGENTO-TWO-30870.zip](https://github.com/magento/magento2/files/5520288/MAGENTO-TWO-30870.zip)

Confirmed this patch fixes the issue on 2.4.3-p1.
